src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+    //  production we need a new linearizable solution.
+    async fn current_upper(&mut self) -> Timestamp {
+        self.write_handle
+            .fetch_recent_upper()
+            .await
+            .as_option()
+            .cloned()
+            .expect("we use a totally ordered time and never finalize the shard")
+    }
+
+    /// Reports if the catalog state has been initialized, and the current upper.
+    async fn is_initialized_inner(&mut self) -> (bool, Timestamp) {
+        let upper = self.current_upper().await;
+        (upper > 0, upper)
+    }
+
+    /// Generates an iterator of [`StateUpdate`] that contain all updates to the catalog
+    /// state up to, but not including, `upper`.
+    ///
+    /// The output is consolidated and sorted by timestamp in ascending order.
+    async fn snapshot(
+        &mut self,
+        upper: Timestamp,
+    ) -> impl Iterator<Item = StateUpdate> + DoubleEndedIterator {
+        let snapshot = if upper > 0 {
+            let since = self.read_handle.since().clone();
+            let mut as_of = upper.saturating_sub(1);
+            as_of.advance_by(since.borrow());
+            self.read_handle
+                .snapshot_and_fetch(Antichain::from_elem(as_of))
+                .await
+                .expect("we have advanced the restart_as_of by the since")
+        } else {
+            Vec::new()
+        };
+        soft_assert!(
+            snapshot.iter().all(|(_, _, diff)| *diff == 1),
+            "snapshot_and_fetch guarantees a consolidated result"
+        );
+        snapshot
+            .into_iter()
+            .map(|((key, _unit), ts, diff)| StateUpdate {
+                kind: key.expect("key decoding error"),
+                ts,
+                diff,
+            })
+            .sorted_by(|a, b| Ord::cmp(&b.ts, &a.ts))
+    }
+
+    /// Get value of config `key`.
+    ///
+    /// Some configs need to be read before the catalog is opened for bootstrapping.
+    async fn get_config(&mut self, key: &str, upper: Timestamp) -> Option<u64> {
+        self.snapshot(upper).await.rev().find_map(
+            |StateUpdate {
+                 kind,
+                 ts: _,
+                 diff: _,
+             }| match kind {
+                StateUpdateKind::Config(k, v) if k.key == key => Some(v.value),
+                _ => None,
+            },
+        )
+    }
+
+    /// Get the user version of this instance.
+    ///
+    /// The user version is used to determine if a migration is needed.
+    async fn get_user_version(&mut self, upper: Timestamp) -> Option<u64> {
+        self.get_config(USER_VERSION_KEY, upper).await
+    }
+
+    /// Politely expires all handles to persist, releasing their leases.
+    // TODO(jkosh44) Before enabling this in production we need to expose this through the trait.
+    async fn _expire(self) {
+        self.read_handle.expire().await;
+        self.write_handle.expire().await;
+    }
```

Comment:
@danhhz I'm not really sure what the correct way of dealing with this is. This struct ideally isn't dropped until all of Materialize is shutting down. Any suggestions?

Reply:
We don't really have clean process shutdown in mz, so not much you can do. We already rely on persist expiring read leases 15m after we deploy, so probably just skip doing anything here?

Reply:
I realized that there is one place where we should call this. When starting up, we sometimes open the catalog in savepoint mode to test that upgrades will work. Then we close it and re-open it in a writeable mode. After closing the savepoint version we should call this method to clean up the handles.

I'll do that in a follow-up PR.